packages/instrumentation-llamaindex/src/instrumentation.ts

```diff
@@ -0,0 +1,160 @@
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type * as llamaindex from "llamaindex";
+
+import {
+  InstrumentationBase,
+  InstrumentationModuleDefinition,
+  InstrumentationNodeModuleDefinition,
+} from "@opentelemetry/instrumentation";
+
+import { LlamaIndexInstrumentationConfig } from "./types";
+import { CustomLLMInstrumentation } from "./custom-llm-instrumentation";
+import { genericWrapper } from "./utils";
+
+import { BaseEmbedding, BaseSynthesizer, LLM, BaseRetriever } from "llamaindex";
+
+export class LlamaIndexInstrumentation extends InstrumentationBase<any> {
+  protected override _config!: LlamaIndexInstrumentationConfig;
+
+  constructor(config: LlamaIndexInstrumentationConfig = {}) {
+    super("@traceloop/instrumentation-llamaindex", "0.0.17", config);
+  }
+
+  public override setConfig(config: LlamaIndexInstrumentationConfig = {}) {
+    super.setConfig(config);
+  }
+
+  public manuallyInstrument(
+    module: typeof llamaindex & { openLLMetryPatched?: boolean },
+  ) {
+    if (module.openLLMetryPatched) {
+      return;
+    }
+  }
+
+  protected init(): InstrumentationModuleDefinition<any> {
+    const module = new InstrumentationNodeModuleDefinition<any>(
+      "llamaindex",
+      [">=0.0.40"],
+      this.patch.bind(this),
+      this.unpatch.bind(this),
+    );
+    return module;
+  }
+
+  private isLLM(llm: any): llm is LLM {
+    return (
+      llm &&
+      (llm as LLM).complete !== undefined &&
+      (llm as LLM).chat !== undefined
+    );
+  }
+
+  private isEmbedding(embedding: any): embedding is BaseEmbedding {
+    return (
+      embedding instanceof BaseEmbedding &&
+      embedding.getQueryEmbedding !== undefined
+    );
+  }
+
+  private isSynthesizer(synthesizer: any): synthesizer is BaseSynthesizer {
+    return (
+      synthesizer && (synthesizer as BaseSynthesizer).synthesize !== undefined
+    );
+  }
+
+  private isRetriever(retriever: any): retriever is BaseRetriever {
+    return retriever && (retriever as BaseRetriever).retrieve !== undefined;
+  }
+
+  private patch(
+    moduleExports: typeof llamaindex & { openLLMetryPatched?: boolean },
+  ) {
+    if (moduleExports.openLLMetryPatched) {
+      return moduleExports;
+    }
+
+    const customLLMInstrumentation = new CustomLLMInstrumentation(
+      this._config,
+      this.tracer,
+    );
+
+    this._wrap(
+      moduleExports.RetrieverQueryEngine.prototype,
+      "query",
+      genericWrapper("query", this.tracer),
+    );
+
+    for (const key in moduleExports) {
+      const cls = (moduleExports as any)[key];
+      if (this.isLLM(cls.prototype)) {
```

Comment:
This one will result in duplicate spans. 

_Explanation:_
We instrument OpenAI (for example) independently of LlamaIndex. It means that for a call to OpenAI, since LlamaIndex uses OpenAI sdk theirselves, you'll have duplicate spans, One reported from the LlamaIndex Instrumentation and the other one from OpenAI instrumentation.

By the naming you gave of `CustomLLMInstrumentation` I guess you saw my python impl.
How I solved this kind of issue there?

In python LllamaIndex have a class called CustomLLM which is intended to be a class where the non trivial/commercial LLMs (openai, cohere, anthropic, etc.) inherits from - example for such one is Ollama.
So I instrumented only those inherit CustomLLM, than the commercial ones which inherit LLM don't get instrumented by LlamaIndex and therefore will not produce span here, but rather produce span through their designated instrumentation.

Unfortunately, I checked, and it seems like there isn't such CustomLLM base class in LlamaIndexTS (validate me).

Maybe we should have some exclusion list of LLMs which have their own instrumentation.

Reply:
But here only calls instrumented in llamaindex are getting captured.

<img width="1792" alt="Screenshot 2024-01-10 at 12 33 11 AM" src="https://github.com/traceloop/openllmetry-js/assets/24036721/b466b6e3-6a50-40b2-bff8-e9b48e4c3d4a">

Code I used for testing
```typescript
import * as traceloop from "@traceloop/node-server-sdk";
import { OpenAI } from "llamaindex";

traceloop.initialize({
  appName: "sample_llamaindex",
  apiKey: process.env.TRACELOOP_API_KEY,
  disableBatch: true,
});

class SampleLlamaIndex {
  @traceloop.workflow("sample_query")
  async query() {
    const openai = new OpenAI();
    const res = await openai.complete('How are you?');
    return res;
  }
}

traceloop.withAssociationProperties(
  { user_id: "12345", chat_id: "789" },
  async () => {
    const sampleLlamaIndex = new SampleLlamaIndex();
    const result = await sampleLlamaIndex.query();
    console.log(result);
  },
);
```